### DISCUSSION TITLE
Parameterized services?

### DISCUSSION THREAD

**Discussion by rfratto** [Homebrew/discussions#4328] [Tap maintenance and Homebrew development] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.6-124-g96ae878
ORIGIN: https://github.com/Homebrew/brew
HEAD: 96ae8788081c42c25f25fecd7723a03ee0f064fd
Last commit: 19 minutes ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 9a6a815e5ea0169b46e10f411ac3237740a6bbdf
Core tap last commit: 4 weeks ago
Core tap branch: master
Core tap JSON: 15 Mar 22:34 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: nvim
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.39.2 => /opt/homebrew/bin/git
Curl: 7.85.0 => /usr/bin/curl
macOS: 13.1-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python@3.11

Warning: Some taps are not on the default git origin branch and may not receive
updates. If this is a surprise to you, check out the default branch with:
  git -C $(brew --repo homebrew/core) checkout master

Warning: Some installed formulae are missing dependencies.
You should `brew install` the missing dependencies:
  brew install kubernetes-cli

Run `brew missing` for more details.
```

### Description of issue

I'm one of the maintainers of [Grafana Agent](https://github.com/grafana/agent), and I'm looking for a way to help users who have installed Grafana Agent via the grafana-agent formula to use one of our new features.

Back in October, we introduced a new runtime mode of the binary where it is enabled once a specific environment variable is set (`AGENT_MODE=flow`). When this new runtime mode is enabled, you are given a different CLI from the normal runtime mode, so you must provide different commands and flags. 

It's not clear how to me how we can help Grafana Agent users using the Homebrew package to enable and configure this mode; the [service generated by the formula](https://github.com/Homebrew/homebrew-core/blob/0384ddd0d2f7868478b393d84ae1abe91e3aed63/Formula/grafana-agent.rb#L58-L63) assumes the old runtime mode, with the old CLI and command line flags. 

My local workaround (on macOS) has been to overwrite the generated plist every time I upgrade the package with the environment variable and custom flags, but this is obviously not a great solution for users. This is less of a problem on Linux, where it's easy to create systemd overrides. 

Is there anything we can do for the formula here? Is it possible to parameterize the generated macOS plist? Or maybe we should add custom flags when the formula is being installed to generate a plist/service definition for that mode?

**Comment by SMillerDev** [comment, marked as answer]
The service block is the single source of truth. We can review a pull request to convert it to the new style, but then everyone will use the new style.

**Reply by rfratto** [reply]
Thanks. Is there nothing we can do at all to allow people to choose?

**Reply by SMillerDev** [reply]
Well, you could always make some pull requests to brew to allow multiple service blocks, figure out a way to make `brew services` deal with multiple blocks, send pull requests to implement that, wait for a brew release to incorporate all of this and make a pull request to add it to the formula. But personally I'd just force everyone to the new version by changing the main block.

**Reply by rfratto** [reply]
Is it not possible to use [options](https://rubydoc.brew.sh/Formula.html#option-class_method) and read the value when generating the service? My Ruby skills aren't good enough to be able to track down what is/isn't possible. At this point we're considering a separate Formula in our tap until we're ready for everyone to migrate over. Forcing a subset of users isn't an option for us right now.

**Reply by gromgit** [reply]
It's probably doable. See [this PR on a third-party repo](https://github.com/shivammathur/homebrew-php/pull/2/files) for an example of how options can be used in formulae.

However, `--with-XXX` options have been _banned_ in Homebrew core repos for a long time now. If options are a must, you'd have to host those formulae in your own tap.
